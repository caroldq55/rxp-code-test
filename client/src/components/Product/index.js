import React from 'react';
import PropTypes from 'prop-types';
import { Button, Col, OverlayTrigger, Row, Tooltip } from 'react-bootstrap';

import {
  ProductContainer,
  ProductImage,
  ProductInfo,
  ProductPrice,
  ProductName,
  ButtonWrapper
} from './styles';

const Product = props => {
  const { product, handleAddToCart } = props;

  const { productImage, productName, price } = product;

  return (
    <ProductContainer>
      <ProductImage src={productImage} alt={productImage} />
      <ProductInfo>
        <Row>
          <Col xs={12} md={8}>
            <OverlayTrigger
              placement='right'
              overlay={<Tooltip>{productName}</Tooltip>}
            >
              <ProductName>{productName}</ProductName>
            </OverlayTrigger>
          </Col>
          <Col xs={12} md={4}>
            <ProductPrice>${price}</ProductPrice>
          </Col>
        </Row>
        <ButtonWrapper>
          <Button
            variant='info'
            onClick={() => {
              handleAddToCart(product);
            }}
            data-test-id='addToCartButton'
          >
            Add to cart
          </Button>
        </ButtonWrapper>
      </ProductInfo>
    </ProductContainer>
  );
};

Product.propTypes = {
  product: PropTypes.shape({
    productImage: PropTypes.string.isRequired,
    productName: PropTypes.string.isRequired,
    price: PropTypes.number.isRequired,
    isPublished: PropTypes.bool
  }).isRequired,
  handleAddToCart: PropTypes.func.isRequired
};

export default Product;
