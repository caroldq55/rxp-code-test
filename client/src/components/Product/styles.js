import styled from 'styled-components';

const ProductContainer = styled.div`
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  padding: 20px;
  margin: 5px;
  position: relative;
`;

const ProductImage = styled.img`
  width: 180px;
`;

const ProductInfo = styled.div`
  margin-top: 10px;
  font-weight: 500;
`;

const ProductName = styled.div`
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
  cursor: pointer;
`;

const ProductPrice = styled.div`
  font-weight: bold;
  color: #333;
`;

const ButtonWrapper = styled.div`
  text-align: center;
  margin-top: 5px;
`;

export {
  ProductContainer,
  ProductImage,
  ProductInfo,
  ProductPrice,
  ProductName,
  ButtonWrapper
};
