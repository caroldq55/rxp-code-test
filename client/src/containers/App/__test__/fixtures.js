import { fromJS } from 'immutable';

const products = [
  {
    isPublished: 'true',
    productName: 'Apple iPhone X',
    productImage:
      'https://www.telstra.com.au/content/dam/tcom/personal/mobile-phones/product-catalogue/iphone-x/iphone-x-silver-grid.png',
    price: '299'
  },
  {
    isPublished: 'true',
    productName: 'Apple iPhone 8',
    productImage:
      'https://www.telstra.com.au/content/dam/tcom/personal/mobile-phones/product-catalogue/iphone-8/iphone-8-silver-grid.png',
    price: '100'
  },
  {
    isPublished: 'false',
    productName: 'Apple iPhone 8 Plus',
    productImage:
      'https://www.telstra.com.au/content/dam/tcom/personal/mobile-phones/product-catalogue/iphone-8/iphone-8plus-space-grey-grid.png',
    price: '99'
  },
  {
    isPublished: 'true',
    productName: 'Sony Xperia XA2',
    productImage:
      'https://www.telstra.com.au/content/dam/tcom/personal/mobile-phones/product-catalogue/sony-xperia-xa2/sony_xperia_xa2_front_v1.png',
    price: '19'
  }
];

const item = {
  isPublished: true,
  productName: 'Sony Xperia XA2',
  productImage:
    'https://www.telstra.com.au/content/dam/tcom/personal/mobile-phones/product-catalogue/sony-xperia-xa2/sony_xperia_xa2_front_v1.png',
  price: '19'
};

const cartItems = [
  {
    isPublished: true,
    productName: 'Sony Xperia XA2',
    productImage:
      'https://www.telstra.com.au/content/dam/tcom/personal/mobile-phones/product-catalogue/sony-xperia-xa2/sony_xperia_xa2_front_v1.png',
    price: '19'
  }
];

const publishedProducts = [
  {
    isPublished: true,
    productName: 'Apple iPhone X',
    productImage:
      'https://www.telstra.com.au/content/dam/tcom/personal/mobile-phones/product-catalogue/iphone-x/iphone-x-silver-grid.png',
    price: '299'
  },
  {
    isPublished: true,
    productName: 'Apple iPhone 8',
    productImage:
      'https://www.telstra.com.au/content/dam/tcom/personal/mobile-phones/product-catalogue/iphone-8/iphone-8-silver-grid.png',
    price: '100'
  },
  {
    isPublished: true,
    productName: 'Sony Xperia XA2',
    productImage:
      'https://www.telstra.com.au/content/dam/tcom/personal/mobile-phones/product-catalogue/sony-xperia-xa2/sony_xperia_xa2_front_v1.png',
    price: '19'
  }
];

const availableProducts = [
  {
    isPublished: true,
    productName: 'Apple iPhone X',
    productImage:
      'https://www.telstra.com.au/content/dam/tcom/personal/mobile-phones/product-catalogue/iphone-x/iphone-x-silver-grid.png',
    price: '299'
  },
  {
    isPublished: true,
    productName: 'Apple iPhone 8',
    productImage:
      'https://www.telstra.com.au/content/dam/tcom/personal/mobile-phones/product-catalogue/iphone-8/iphone-8-silver-grid.png',
    price: '100'
  }
];

const initialValue = fromJS({
  products: [],
  loading: false,
  cartItems: []
});

export {
  products,
  initialValue,
  cartItems,
  availableProducts,
  item,
  publishedProducts
};
