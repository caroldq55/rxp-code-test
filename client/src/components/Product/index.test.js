import React from 'react';
import { shallow } from 'enzyme';
import 'jest-styled-components';

import Product from '.';
import { item } from '../../containers/App/__test__/fixtures';

describe('<Product/>', () => {
  const handleAddToCartSpy = jest.fn();

  it('should render Product correctly', () => {
    const wrapper = shallow(
      <Product product={item} handleAddToCart={handleAddToCartSpy} />
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('should render Product correctly', () => {
    const wrapper = shallow(
      <Product product={item} handleAddToCart={handleAddToCartSpy} />
    );
    const addToCartButton = wrapper.find('[data-test-id="addToCartButton"]');
    expect(addToCartButton).toHaveLength(1);
    addToCartButton.simulate('click');
    expect(handleAddToCartSpy).toHaveBeenCalledTimes(1);
  });
});
