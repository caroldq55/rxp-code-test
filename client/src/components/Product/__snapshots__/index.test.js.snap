// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Product/> should render Product correctly 1`] = `ShallowWrapper {}`;
